### PR TITLE
#163519153 Implemented search functionality and custom filtering

### PR DESCRIPTION
**Pull request by filschristian**
#### What does this PR do?
Have API end-points for searching articles with the capability of custom filtering

#### Description of Task to be completed?
Have the following API end-point working:
`GET: api/v1/articles/search?title=new article&author=chris&tag=java`

#### How should this be manually tested?

- Pull this branch

- Run `npm run migrate:seed`

- Run `npm run test`

- Run `npm run dev`

- With postman send post requests to that API end-point with parameters you want to use filtering

#### Any background context you want to provide?
No Authentication required for this API end-points.

#### What are the relevant pivotal tracker stories?
#163519153

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

File: controllers/ArticleController.js [hunk 1]
```diff
@@ -214,6 +214,56 @@ class ArticleController {
...
+    }
+    const articles = await Article.findAndCountAll({
+      where,
+      include,
+      limit,
+      offset
+    });
+    pages = Math.ceil(articles.count / limit);
+    if (articles.length <= 0) {
+      return res.status(404).json({ status: 404, message: 'Nothing found' });
```
Comment (NkFab) [inline controllers/ArticleController.js]: @filschristian this line is not covered in tests. Try to cover it.

File: __tests__/routes/article.test.js [hunk 2]
```diff
@@ -54,7 +62,7 @@ describe('articles', () => {
       }
     }).then(() => true);
     await Article.destroy({
-      where: { tagList: { [Op.contains]: ['Test'] } }
+      where: { tagList: { [Op.contains]: ['test'] } }
```
Comment (NkFab) [inline __tests__/routes/article.test.js]: @filschristian  Why this change? from 'Test' to 'test'. Wasn't it working prior to this change?

Comment (filschristian) [reply __tests__/routes/article.test.js]: > @filschristian Why this change? from 'Test' to 'test'. Wasn't it working prior to this change?

Yes, It was not working because of the capital T

File: controllers/ArticleController.js (hunk 1 above)
Comment (filschristian) [reply controllers/ArticleController.js]: > @filschristian this line is not covered in tests. Try to cover it.

Feedback implemented. Thank you for the feedback

File: __tests__/routes/article.test.js (hunk 2 above)
Comment (NkFab) [reply __tests__/routes/article.test.js]: Ohh I see

**Review by NkFab — APPROVED** [review]

File: __tests__/routes/article.test.js [hunk 3]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
     expect(res.body.message).toBe('Article deleted successfully');
   });
+
+  test('Delete - should return article not found', async () => {
+    expect.assertions(3);
+    const res = await request(app)
+      .delete(`${urlPrefix}/articles/${fakeSlug}`)
+      .set('Authorization', loginUser1.token);
+    expect(res.status).toBe(404);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 4]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.message).toBe('Article not found');
+  });
+
+  test('like an unexisting article', async () => {
+    expect.assertions(3);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${fakeSlug}/like`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(404);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 5]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.message).toBe('Article not found');
+  });
+
+  test('like an article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/like`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(201);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 6]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.article).toBeDefined();
+  });
+
+  test('dislike a liked article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/dislike`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(200);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 7]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.message).toBe('Disliked');
+  });
+
+  test('Remove dislike from an article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/dislike`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(200);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js:248]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 8]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.article).toBeDefined();
+  });
+
+  test('dislike an article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/dislike`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(200);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 9]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.message).toBe('Disliked');
+  });
+
+  test('dislike an unexisting article', async () => {
+    expect.assertions(3);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${fakeSlug}/dislike`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(404);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 10]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.message).toBe('Article not found');
+  });
+
+  test('like a disliked article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/like`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(200);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

File: __tests__/routes/article.test.js [hunk 11]
```diff
@@ -200,4 +204,128 @@ describe('articles', () => {
...
+    expect(res.body.article).toBeDefined();
+  });
+
+  test('Remove like from an article', async () => {
+    expect.assertions(4);
+    const res = await request(app)
+      .post(`${urlPrefix}/articles/${newArticle.slug}/like`)
+      .set('Authorization', loginUser2.token);
+    expect(res.status).toBe(200);
+    expect(res.body).toBeDefined();
```
Comment (Nziranziza) [inline __tests__/routes/article.test.js]: Testing for the body is not necessary because on the next line you are expecting a property of body and if it is true it means that the body also is defined. So I suggest removing it

**Review by Nziranziza — APPROVED** [review]
